network/src/protocols/test.rs

```diff
@@ -212,19 +212,12 @@ fn net_service_start(name: String) -> Node {
     let peer_id = network_state.local_peer_id().clone();
 
     let mut listen_addr = p2p_service
-        .listen("/ip4/0.0.0.0/tcp/0".parse().unwrap())
+        .listen("/ip4/127.0.0.1/tcp/0".parse().unwrap())
```

Comment:
`127.0.0.1` will not trigger this discovery behavior, here must use `0.0.0.0`

https://github.com/nervosnetwork/p2p/blob/master/protocols/discovery/src/substream.rs#L381

Reply:
Unfortunately `0.0.0.0` is definitely also an unreachable "address".

Next step we need move `is_reachable` check from p2p lib to ckb, to allow us write tests.

Reply:
But `0.0.0.0` is not loopback，so it can pass the check，it will pass not loopback or can reachable

Reply:
I see, seems like a bug, why do we want to discovery a "not loopback and unreachable address"?

Also "0.0.0.0" is an "unusesal address", see https://en.wikipedia.org/wiki/0.0.0.0

In fact, even change listen_addr back to `0.0.0.0`, the IP of a peer the node read from socket is high probably still `127.0.0.1`

Reply:
Because the discovery of these two kind addresses is meaningless, it cannot be connected.

`0.0.0.0` is a special address, it will bind all the network cards of the device, and it is impossible to know exactly whether it is valid, so the verification can be let go.

`0.0.0.0`not equivalent to 127.0.0.1

Reply:
Maybe more reasonable conditions should be ` RawAddr::from(address).is_reachable() ||  !address.ip().is_loopback() `

Reply:
Maybe more reasonable conditions should be ` RawAddr::from(address).is_reachable() ||  !address.ip().is_loopback() `

Reply:
IMO the condition should be `RawAddr::from(address).is_reachable() && !address.ip().is_loopback()` or simply `RawAddr::from(address).is_reachable()`.

And this check should be removed from p2p library side, the p2p library can only focus on handle protocols, let ckb handle the logic.

Reply:
> And this check should be removed from p2p library side, the p2p library can only focus on handle protocols, let ckb handle the logic.

I think general inspections can be used in the library. 

> `RawAddr::from(address).is_reachable() && !address.ip().is_loopback()`

This will block `0.0.0.0`

Reply:
> > And this check should be removed from p2p library side, the p2p library can only focus on handle protocols, let ckb handle the logic.
> 
> I think general inspections can be used in the library.
> 
> > `RawAddr::from(address).is_reachable() && !address.ip().is_loopback()`
> 
> This will block `0.0.0.0`

You can't guarantee that other users of this library don't want to discovery internal IP addresses. And most important we can't write tests code to check discovery and identify ...

`RawAddr::from(address).is_reachable() && !address.ip().is_loopback()` is Ok.

Reply:
This is actually the built-in protocol behavior of discovery, which has two means:
1. Interchange stored discovered addresses
2. Timing interactively connecting addresses between nodes

This is the protocol behavior